zerver/tests/test_auth_backends.py

```diff
@@ -1872,6 +1872,364 @@ def test_idp_initiated_signin_subdomain_implicit_invalid(self) -> None:
+
+    LOGIN_URL = "/accounts/login/social/apple"
+    SIGNUP_URL = "/accounts/register/social/apple"
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         desktop_flow_otp: Optional[str]=None,
+                         is_signup: bool=False,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         alternative_start_url: Optional[str]=None,
+                         user_agent: Optional[str]=None,
+                         **extra_data: Any) -> HttpResponse:
+        url, headers = self.prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key, alternative_start_url,
+            user_agent=user_agent,
+        )
+
+        result = self.client_get(url, **headers)
+
+        expected_result_url_prefix = 'http://testserver/login/%s/' % (self.backend.name,)
+        if settings.SOCIAL_AUTH_SUBDOMAIN is not None:
+            expected_result_url_prefix = ('http://%s.testserver/login/%s/' %
+                                          (settings.SOCIAL_AUTH_SUBDOMAIN, self.backend.name,))
+
+        if result.status_code != 302 or not result.url.startswith(expected_result_url_prefix):
+            return result
+
+        result = self.client_get(result.url, **headers)
+        self.assertEqual(result.status_code, 302)
+        assert self.AUTHORIZATION_URL in result.url
+
+        self.client.cookies = result.cookies
+
+        # id_token in a JSON Web Token that contains the user’s identity information.
+        token_data_dict = {
+            'access_token': 'foobar',
+            'expires_in': time.time() + 60*5,
+            'id_token': self.generate_id_token(account_data_dict),
+            'token_type': 'bearer'
+        }
+
+        parsed_url = urllib.parse.urlparse(result.url)
+        state = urllib.parse.parse_qs(parsed_url.query)['state']
+
+        with responses.RequestsMock(assert_all_requests_are_fired=False) as requests_mock:
+            requests_mock.add(
+                requests_mock.POST,
+                self.ACCESS_TOKEN_URL,
+                match_querystring=False,
+                status=200,
+                body=json.dumps(token_data_dict))
+
+            # The below mock is to get public key we generated
+            # instead of getting it from
+            # https://appleid.apple.com/auth/keys.
+            requests_mock.add(
+                requests_mock.GET,
+                self.BACKEND_CLASS.JWK_URL,
+                status=200,
+                json=json.loads(settings.APPLE_JWK),
+            )
+
+            self.client.session.flush()
+            result = self.client_post(self.AUTH_FINISH_URL,
+                                      dict(state=state), **headers)
+        return result
```

Comment:
Do we need to copy all of `social_auth_test`?  That function is designed to be extensible, with e.g. `self.register_extra_endpoints` (which handles at least one of the issues here), and I'd really like to avoid making a bunch of almost-copies of it as that's bad for maintainability.

I think you can achieve what you're trying to do by just using its existing pluggable features and maybe adding a new argument to pass the extra `token_data_dict` data.

Reply:
That looks to be the best way. made use of `social_auth_test_finish` and `register_extra_endpoins` and removed `social_auth_test`